### PR TITLE
test(integration): add test convention

### PR DESCRIPTION
**Pull request by thxCode**
#45

- Refactor some setup register entrances to support configuring in integration tests.
- Introduce [Ginkgo](https://onsi.github.io/ginkgo/) and [Kind](https://kind.sigs.k8s.io/docs/user/quick-start/) as the testing base, and make an integration test framework for the Ginkgo suite entrance.
- Use [Ginkgo](https://onsi.github.io/ginkgo/) syntax to write some cases.

**How to run:**

- If there is an existing fresh cluster, we can locate the KubeConfig to local and run `USE_EXISTING_CLUSTER=true ./script/test-integration` to trigger the integration testing.
- If there is no cluster, we can run `./script/test-integration` or `make test-integration` to create a local [Kind](https://kind.sigs.k8s.io/docs/user/quick-start/) cluster and trigger the integration testing.
- By default, the testing CRs will `Cleanup` after each test case, if we want to review these CRs, we can use `KEEP_TESTING_RESOURCE=true`. **We should never use `KEEP_TESTING_RESOURCE` during a CI phase, it may leak a testing cluster.**

**Not the goal:***

- Launch the local cluster via [K3D](https://github.com/rancher/k3d).
- Specify custom KubeConfig in rancher dapper, so we cannot use `USE_EXISTING_CLUSTER=true make test-integration`.

### CODE REVIEW

**Comment by thxCode** [comment]
I will squash the `refactor` commits and `test` commits after LGTMed, cc @guangbochen.

File: test/framework/envtest/cluster/finder/env.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"time"
+)
+
+type EnvFinder struct {
+	prefix             string
+	punctuationPattern *regexp.Regexp
+}
+
+func (f *EnvFinder) Get(camelEnvName string, defaultVal string) string {
+	var envName = f.punctuationPattern.ReplaceAllStringFunc(camelEnvName, func(s string) string {
```
Comment (orangedeng) [inline test/framework/envtest/cluster/finder/env.go]: I am not sure you are going to convert a camel case string to snake case. If so, I think we can use `https://github.com/iancoleman/strcase` like https://play.golang.org/p/_-HChDbi2Vd

**Comment by gitlawr** [comment]
Are we setting up the kind cluster twice, one for master and one for global? Can we do it once

File: test/framework/envtest/cluster/finder/env.go (hunk 1 above)
Comment (thxCode) [reply test/framework/envtest/cluster/finder/env.go]: Updated

**Comment by thxCode** [comment]
@gitlawr , setting up an independence cluster for each suite is a reasonable way, at the same time, testing in an existing cluster is also supported. It seems we can only consider the steve API in the integration test, so I merge `global` and `master` suites in one, but the principle remains the same. It is recommended that each independent suite be run in an independent environment.

File: test/integration/api/suite_test.go [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+	// which means the cache(informer) has not ready,
+	// so we give a stupid time sleep to trigger the first list-watch,
+	// and please use the client interface instead of informer interface if you can.
+	select {
+	case <-time.After(15 * time.Second):
+	case err := <-testSuiteStartErrChan:
+		// release the chan
+		close(testSuiteStartErrChan)
+		testSuiteStartErrChan = nil
+		GinkgoT().Errorf("failed to start harvester server, %v", err)
```
Comment (gitlawr) [inline test/integration/api/suite_test.go:107]: Are we logging an error when `harvester.Start()` returns nil?
There are some principals for channels:
We usually don't need to close them
A receiver should never close them

File: test/integration/api/suite_test.go [hunk 3]
```diff
@@ -0,0 +1,648 @@
...
+
+	// create docker client
+	dClient, err := dockerclient.NewClientWithOpts(dockerclient.FromEnv)
+	if err != nil {
+		berr = fmt.Errorf("failed to create docker client, %v", err)
+		return
+	}
+
+	// create minio container
+	minoCreatedResp, err := dClient.ContainerCreate(
```
Comment (gitlawr) [inline test/integration/api/suite_test.go:499]: Why don't we install minio in kubernetes way like the others?

**Comment by guangbochen** [comment]
close in favor of https://github.com/rancher/harvester/pull/258

**Pull request closed without merging**